README.md
About
=====

This is the platform pattern designed to build the system the WordPress service.
Currently supported:

* CentOS (6.5 or later and 7.x)

For more information, please visit [official web site](http://cloudconductor.org/).

Requirements
============

Prerequisites
-------------

- cloudconductor (>= 2.0)

How to use patterns
============

You can apply this pattern by using CloudConductor CLI tools or REST API.
Please see [Getting started](http://cloudconductor.org/) in official web site to know
how the parameters you input to CloudConductor CLI tools or REST API are converted into
chef attributes.

Attributes
==========

The attributes not described here have default values, and you can change the value of them if you need.
Please see the attribute files if you want to know what kind of attributes are in this pattern.


Copyright and License
=====================

Copyright 2016 TIS inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


Contact
========

For more information: <http://cloudconductor.org/>

Report issues and requests: <https://github.com/cloudconductor-patterns/simple_wordpress_pattern/issues>

Send feedback to: <ccndctr@gmail.com>


CHANGELOG.md
CHANGELOG
=========

## version 1.0.0 (2016/03/30)

  -  First release of this pattern that contains wordpress


test/scripts/setup.sh
#!/bin/sh -ex

cd $(dirname $0)/data

source ./scripts/setup_wordpress.sh wordpress 2>&1


test/scripts/configure.sh
#!/bin/sh -ex

cd $(dirname $0)/data

source ./scripts/setup_wordpress.sh wordpress 2>&1

if [ ! -d /opt/cloudconductor ]; then
  mkdir -p /opt/cloudconductor
fi

cat <<_EOF_ > /opt/cloudconductor/cfn_parameters
WordPressTitle="example"
WordPressAdminUser="admin"
WordPressAdminPassword="password"
WordPressAdminEmail="admin@example.com"
_EOF_

which systemctl && systemctl start httpd ||  service httpd start

source ./scripts/configure_wordpress.sh wordpress 2>&1

source ./scripts/spec_wordpress.sh wordpress 2>&1


lib/events.sh
#!/bin/sh

function setup_wordpress() {
  local dbname=mysql
  local dbsvcname=mysqld
  if [ "$(os_version)" == "7" ]; then
    dbname=mariadb
    dbsvcname=mariadb
  fi

  # Install dependencies
  yum -y update
  yum -y install python-setuptools
  easy_install https://s3.amazonaws.com/cloudformation-examples/aws-cfn-bootstrap-latest.tar.gz
  yum -y install httpd ${dbname}-server
  yum -y install php php-mysql php-mbstring php-gd
  service_ctl enable httpd
  service_ctl enable ${dbsvcname}

  # Create Database User
  service_ctl start ${dbsvcname}
  if [ ! -f /var/lib/mysql/mysql.sock ]; then
    touch /var/lib/mysql/mysql.sock
    chown mysql:mysql /var/lib/mysql
    service_ctl restart ${dbsvcname}
  fi

  if [ -f ~/.wordpress_mysql_password ]; then
    wordpress_mysql_password=`cat ~/.wordpress_mysql_password`
  else
    wordpress_mysql_password=`head -c16 /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 16 | head -n 1`
    echo $wordpress_mysql_password > ~/.wordpress_mysql_password
  fi
  mysql -u root -e "CREATE DATABASE wordpress CHARACTER SET utf8;"
  mysql -u root -e "GRANT ALL PRIVILEGES ON wordpress.* TO wordpress@localhost IDENTIFIED BY '${wordpress_mysql_password}';"
  mysql -u root -e "FLUSH PRIVILEGES;"

  # Download WordPress
  curl -o /usr/local/bin/wp https://raw.githubusercontent.com/wp-cli/builds/gh-pages/phar/wp-cli.phar
  chmod a+x /usr/local/bin/wp
  [ -d /var/www/wordpress ] || mkdir /var/www/wordpress
  chown apache:apache /var/www/wordpress
  sudo -u apache -- /usr/local/bin/wp core download --path=/var/www/wordpress --locale=ja --version=4.4
  sudo -u apache -- /usr/local/bin/wp core config --path=/var/www/wordpress \
    --dbname=wordpress --dbuser=wordpress --dbpass="${wordpress_mysql_password}"
}

function wordpress_first_settings() {
  # Configure WordPress Settings
  load_cfn_parameters
  sudo -u apache -- /usr/local/bin/wp core install --path=/var/www/wordpress --url="${WordPressUrl}" \
    --title="'${WordPressTitle}'" --admin_user="${WordPressAdminUser}" \
    --admin_password="${WordPressAdminPassword}" --admin_email="${WordPressAdminEmail}"
  sudo -u apache -- /usr/local/bin/wp core update --path=/var/www/wordpress
  sudo -u apache -- /usr/local/bin/wp core update-db --path=/var/www/wordpress
  sudo -u apache -- /usr/local/bin/wp plugin update --all --path=/var/www/wordpress
}

function get_domain() {
  load_cfn_parameters
  if [ -n "${WordPressUrl}" ]; then
    url="${WordPressUrl}"
    proto=`echo "${WordPressUrl}" | sed -e 's,^\(.*://\).*,\1,g'`
    url=`echo "${WordPressUrl/$proto/}"`
    host=`echo "$url" | cut -d/ -f1`
    echo -n "$host"
  else
    self_address=`curl -s "http://169.254.169.254/latest/meta-data/public-ipv4"`
    echo -n "$self_address"
  fi
}

function update_apache_virtual_host() {
  server_name=`get_domain`
  if [ -n "${server_name}" ]; then
    cat > /etc/httpd/conf.d/wordpress.conf <<-EOF
<VirtualHost *:80>
  DocumentRoot "/var/www/wordpress"
  ServerName "${server_name}"
</VirtualHost>
EOF
    service_ctl status httpd
    if [ $? -eq 0 ]; then
      service_ctl reload httpd
    else
      service_ctl restart httpd
    fi
  fi
}

function configure_wordpress() {
  if [ ! -f /tmp/.wordpress_installed ]; then
    wordpress_first_settings
    touch /tmp/.wordpress_installed
  fi
  update_apache_virtual_host
}

function spec_wordpress() {
  # it should response 200 OK
  status_code=`curl -sLI http://localhost/ --noproxy localhost -o /dev/null -w '%{http_code}\n'`
  if [ "${status_code}" != "200" ]; then
    echo "localhost:80 returns ${status_code}" 1>&2
    curl http://localhost/ --noproxy localhost
    exit 1
  fi
}


scripts/functions.sh
#!/bin/sh

function consul_kv_get() {
  # $1: consul key
  usage="Usage: consul_kv_get KEY"
  key=${1:?missing key. $usage}
  secret_key=`echo $CONSUL_SECRET_KEY | sed -e 's/\+/%2b/g'`
  consul_url="http://localhost:8500/v1/kv/${key}?token=${secret_key}"
  curl --insecure -s $consul_url | jq .[].Value | sed s/\"//g | base64 -d | jq .;
}

function consul_kv_set() {
  # $1: consul key
  # $2: new value
  usage="Usage: consul_kv_set KEY NEW_VALUE"
  key=${1:?missing key. $usage}
  value=${2:?missing value. $usage}
  secret_key=`echo $CONSUL_SECRET_KEY | sed -e 's/\+/%2b/g'`
  consul_url="http://localhost:8500/v1/kv/${key}?token=${secret_key}"
  curl --insecure -s -X PUT "${consul_url}" -d "${value}"
}

function load_cfn_parameters() {
  [ -f /opt/cloudconductor/cfn_parameters ] && source /opt/cloudconductor/cfn_parameters
}

function get_consul_parameters() {
  # no arguments
  consul_kv_get "cloudconductor/parameters"
}

function run() {
  local e E T oldIFS
  [[ ! "$-" =~ e ]] || e=1
  [[ ! "$-" =~ E ]] || E=1
  [[ ! "$-" =~ T ]] || T=1

  set +e
  set +E
  set +T

  output="$("$@" 2>&1)"
  status="$?"
  oldIFS=$IFS
  IFS=$'\n' lines=($output)

  IFS=$oldIFS
  [ -z "$e" ] || set -e
  [ -z "$E" ] || set -E
  [ -z "$T" ] || set -T
}

function os_version() {
  if [ -f /etc/redhat-release ]; then
    rpm -qf --queryformat="%{VERSION}" /etc/redhat-release
  fi
}

function service_ctl() {
  local action=$1
  local svcname=$2

  local os_version=$(os_version)

  case ${os_version} in
    '6' )
      service_ctl_el6 ${action} ${svcname} || return $?
      ;;
    '7')
      service_ctl_el7 ${action} ${svcname} || return $?
      ;;
  esac
}

function service_ctl_el6() {
  local action=$1
  local svcname=$2
  run bash -c "service --status-all | grep ${svcname}"
  if [ $status -eq 0 ]; then
    case ${action} in
      'enable' )
        /sbin/chkconfig --add ${svcname}
        /sbin/chkconfig ${svcname} on || return $?
        ;;
      'disable' )
        /sbin/chkconfig ${svcname} off
        service ${svcname} stop
        ;;
      *)
        service ${svcname} ${action} || return $?
        ;;
    esac
  fi
}

function service_ctl_el7() {
  local action=$1
  local svcname=$2

  run bash -c "systemctl --all | grep \"^ *${svcname}\""
  if [ "$status" -eq 0 ]; then
    case ${action} in
      'disable' )
        systemctl stop ${svcname}
        systemctl disable ${svcname}
        ;;
      *)
        systemctl ${action} ${svcname} || return $?
        ;;
    esac
  else
    case ${action} in
      'enable' )
        systemctl enable ${svcname} #|| return $?
        ;;
    esac
  fi
}


scripts/configure_wordpress.sh
#!/bin/sh -e
set -o pipefail
[ -f scripts/functions.sh ] && source scripts/functions.sh

function get_domain() {
  domain=`consul_kv_get "cloudconductor/system_domain" | grep -o '"dns": ".*"' | cut -d '"' -f4`
  if [ -n "${domain}" ]; then
    echo -n "${domain}"
  else
    self_address=`curl -s "http://169.254.169.254/latest/meta-data/public-ipv4"`
    echo -n "$self_address"
  fi
}

function wordpress_first_settings() {
  # Configure WordPress Settings
  load_cfn_parameters
  wordpress_domain=`get_domain`
  sudo -u apache -- /usr/local/bin/wp core install --path=/var/www/wordpress --url="http://${wordpress_domain}" \
    --title="'${WordPressTitle}'" --admin_user="${WordPressAdminUser}" \
    --admin_password="${WordPressAdminPassword}" --admin_email="${WordPressAdminEmail}"
  sudo -u apache -- /usr/local/bin/wp core update --path=/var/www/wordpress
  sudo -u apache -- /usr/local/bin/wp core update-db --path=/var/www/wordpress
  sudo -u apache -- /usr/local/bin/wp plugin update --all --path=/var/www/wordpress
}

function update_apache_virtual_host() {
  server_name=`get_domain`
  if [ -n "${server_name}" ]; then
    cat > /etc/httpd/conf.d/wordpress.conf <<-EOF
<VirtualHost *:80>
  DocumentRoot "/var/www/wordpress"
  ServerName "${server_name}"
</VirtualHost>
EOF
    service_ctl status httpd
    if [ $? -eq 0 ]; then
      service_ctl reload httpd
    else
      service_ctl restart httpd
    fi
  fi
}

function configure_wordpress() {
  if [ ! -f /tmp/.wordpress_installed ]; then
    wordpress_first_settings
    touch /tmp/.wordpress_installed
  fi
  update_apache_virtual_host
}

role=$1

if [ "${role}" == "wordpress" ]; then
  configure_wordpress
else
  echo "Unsupported role '${role}'."
fi


scripts/setup_wordpress.sh
#!/bin/sh -e
set -o pipefail
[ -f scripts/functions.sh ] && source scripts/functions.sh

function setup_wordpress() {
  local dbname=mysql
  local dbsvcname=mysqld
  if [ "$(os_version)" == "7" ]; then
    dbname=mariadb
    dbsvcname=mariadb
  fi

  # Install dependencies
  yum -y update
  yum -y install httpd ${dbname}-server
  yum -y install php php-mysql php-mbstring php-gd
  service_ctl enable httpd
  service_ctl enable ${dbsvcname}

  run which cfn-get-metadata
  if [ $status -ne 0 ] ; then
    yum -y install python-setuptools
    easy_install https://s3.amazonaws.com/cloudformation-examples/aws-cfn-bootstrap-latest.tar.gz
  fi

  # Create Database User
  service_ctl start ${dbsvcname}
  if [ ! -f /var/lib/mysql/mysql.sock ]; then
    touch /var/lib/mysql/mysql.sock
    chown mysql:mysql /var/lib/mysql
    service_ctl restart ${dbsvcname}
  fi

  if [ -f ~/.wordpress_mysql_password ]; then
    wordpress_mysql_password=`cat ~/.wordpress_mysql_password`
  else
    wordpress_mysql_password=`head -c16 /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 16 | head -n 1`
    echo $wordpress_mysql_password > ~/.wordpress_mysql_password
  fi
  mysql -u root -e "CREATE DATABASE wordpress CHARACTER SET utf8;"
  mysql -u root -e "GRANT ALL PRIVILEGES ON wordpress.* TO wordpress@localhost IDENTIFIED BY '${wordpress_mysql_password}';"
  mysql -u root -e "FLUSH PRIVILEGES;"

  # Download WordPress
  curl -o /usr/local/bin/wp https://raw.githubusercontent.com/wp-cli/builds/gh-pages/phar/wp-cli.phar
  chmod a+x /usr/local/bin/wp
  [ -d /var/www/wordpress ] || mkdir /var/www/wordpress
  chown apache:apache /var/www/wordpress
  sudo -u apache -- /usr/local/bin/wp core download --path=/var/www/wordpress --locale=ja --version=4.4
  sudo -u apache -- /usr/local/bin/wp core config --path=/var/www/wordpress \
    --dbname=wordpress --dbuser=wordpress --dbpass="${wordpress_mysql_password}"
}

role=$1

if [ "${role}" == "wordpress" ]; then
  setup_wordpress
else
  echo "Unsupported role '${role}'."
fi
